### PR TITLE
Uses LRU to cache gaussian random number in random_projection

### PR DESCRIPTION
**Pull request by kazuki**
This pull request is based from non-merged branch.

based branch: #260 ~~based branch: #246~~ (updated: 2016/03/28)
related issue: #244

### CODE REVIEW

**Comment by kazuki** [comment]
rebased

File: jubatus/core/nearest_neighbor/euclid_lsh.cpp [hunk 1]
```diff
@@ -124,6 +124,15 @@ void euclid_lsh::set_config(const config& conf) {
   }
   hash_num_ = conf.hash_num;
   threads_ = read_threads_config(conf.threads);
+
+  if (conf.cache_size.bool_test()) {
+    if (!(0 <= *(conf.cache_size))) {
+      throw JUBATUS_EXCEPTION(
+          common::invalid_parameter("0 <= cache_size"));
+    }
+    if (*(conf.cache_size) > 0)
```
Comment (TkrUdagawa) [inline jubatus/core/nearest_neighbor/euclid_lsh.cpp]: coding rule
https://gist.github.com/unnonouno/4655412#braces

File: jubatus/core/nearest_neighbor/lsh.cpp [hunk 2]
```diff
@@ -55,6 +55,15 @@ void lsh::set_config(const config& conf) {
         common::invalid_parameter("1 <= hash_num"));
   }
   threads_ = read_threads_config(conf.threads);
+
+  if (conf.cache_size.bool_test()) {
+    if (!(0 <= *(conf.cache_size))) {
+      throw JUBATUS_EXCEPTION(
+          common::invalid_parameter("0 <= cache_size"));
+    }
+    if (*(conf.cache_size) > 0)
```
Comment (TkrUdagawa) [inline jubatus/core/nearest_neighbor/lsh.cpp]: coding rule. need braces for all if statements.

File: jubatus/core/nearest_neighbor/euclid_lsh.cpp (hunk 1 above)
Comment (kazuki) [inline jubatus/core/nearest_neighbor/euclid_lsh.cpp]: fixed

File: jubatus/core/nearest_neighbor/lsh.cpp (hunk 2 above)
Comment (kazuki) [inline jubatus/core/nearest_neighbor/lsh.cpp]: fixed

File: jubatus/core/nearest_neighbor/lsh_function.cpp [hunk 3]
```diff
@@ -27,46 +27,64 @@
...
+using jubatus::util::concurrent::scoped_lock;
 
 namespace jubatus {
 namespace core {
 namespace nearest_neighbor {
 namespace {
 
 std::vector<float> random_projection_internal(
-  const common::sfv_t& sfv, uint32_t hash_num, size_t start, size_t end);
+    const common::sfv_t& sfv, uint32_t hash_num,
```
Comment (TkrUdagawa) [inline jubatus/core/nearest_neighbor/lsh_function.cpp]: coding rule
https://gist.github.com/unnonouno/4655412#function-declarations-and-definitions

File: jubatus/core/nearest_neighbor/lsh_function.cpp [hunk 4]
```diff
@@ -27,46 +27,64 @@
...
 template <class RND>
-inline void next_gaussian_float8(RND& g, float *out);
+inline void next_gaussian_float8(RND& g, float* out);
 #endif
 
 #ifdef JUBATUS_USE_FMV
 __attribute__((target("default")))
 std::vector<float> random_projection_internal(
-  const common::sfv_t& sfv, uint32_t hash_num, size_t start, size_t end);
+    const common::sfv_t& sfv, uint32_t hash_num,
```
Comment (TkrUdagawa) [inline jubatus/core/nearest_neighbor/lsh_function.cpp]: coding rule
https://gist.github.com/unnonouno/4655412#function-declarations-and-definitions

File: jubatus/core/nearest_neighbor/lsh_function.cpp (hunk 3 above)
Comment (kazuki) [inline jubatus/core/nearest_neighbor/lsh_function.cpp]: fixed

File: jubatus/core/nearest_neighbor/lsh_function.cpp (hunk 4 above)
Comment (kazuki) [inline jubatus/core/nearest_neighbor/lsh_function.cpp]: fixed

File: jubatus/core/nearest_neighbor/lsh_function.cpp [hunk 5]
```diff
@@ -273,6 +357,59 @@ inline static void next_gaussian_float16(RND& g, float *out) {
...
+  }
+}
+
+inline static bool check_cache(
+    cache_t& cache,
+    uint32_t seed,
+    vector<float>& proj,
+    float v) {
+  if (cache.bool_test()) {
+    scoped_lock lk(cache->lock);
```
Comment (TkrUdagawa) [inline jubatus/core/nearest_neighbor/lsh_function.cpp]: I think read lock would be better.

Comment (TkrUdagawa) [inline jubatus/core/nearest_neighbor/lsh_function.cpp]: I see that lru.get is not const function. So, this lock need to be mutex or write lock.

**Comment by TkrUdagawa** [comment]
LGTM!

**Comment by TkrUdagawa** [comment]
We need to add document about cache_size option.

**Pull request merged**